File: services/ai/telemetry.py
import logging
import os
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION, DEPLOYMENT_ENVIRONMENT
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from ulid import ULID

logger = logging.getLogger(__name__)


def init_telemetry(app, service_name: str = "omni-ai"):
    """
    Initialize OpenTelemetry instrumentation for the FastAPI application.
    """
    otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    deployment_id = os.getenv("OTEL_DEPLOYMENT_ID", str(ULID()))
    environment = os.getenv("OTEL_DEPLOYMENT_ENVIRONMENT", "development")
    service_version = os.getenv("SERVICE_VERSION", "0.1.0")

    # Create resource with service information
    resource = Resource(
        attributes={
            SERVICE_NAME: service_name,
            SERVICE_VERSION: service_version,
            DEPLOYMENT_ENVIRONMENT: environment,
            "deployment.id": deployment_id,
        }
    )

    # Create tracer provider
    provider = TracerProvider(resource=resource)

    # Add OTLP exporter if endpoint is configured
    if otlp_endpoint:
        logger.info(f"Initializing OpenTelemetry with OTLP endpoint: {otlp_endpoint}")
        otlp_exporter = OTLPSpanExporter(endpoint=f"{otlp_endpoint}/v1/traces")
        processor = BatchSpanProcessor(otlp_exporter)
        provider.add_span_processor(processor)
    else:
        logger.info("No OTLP endpoint configured, telemetry will be collected locally only")

    # Set the global tracer provider
    trace.set_tracer_provider(provider)

    # Instrument FastAPI
    FastAPIInstrumentor.instrument_app(app)

    # Instrument HTTPX (for outbound HTTP requests)
    HTTPXClientInstrumentor().instrument()

    logger.info(
        f"Telemetry initialized for {service_name} "
        f"(deployment_id={deployment_id}, environment={environment})"
    )


def get_tracer(name: str = "omni-ai"):
    """
    Get a tracer instance for manual instrumentation.
    """
    return trace.get_tracer(name)

File: web/src/lib/server/telemetry.ts
import { NodeSDK } from '@opentelemetry/sdk-node'
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http'
import { Resource } from '@opentelemetry/resources'
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from '@opentelemetry/semantic-conventions'
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node'
import { ulid } from 'ulid'
import { propagation, trace, context, type Span } from '@opentelemetry/api'

let sdk: NodeSDK | null = null

export function initTelemetry() {
    const otlpEndpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT
    const deploymentId = process.env.OTEL_DEPLOYMENT_ID || ulid()
    const environment = process.env.OTEL_DEPLOYMENT_ENVIRONMENT || 'development'
    const serviceVersion = process.env.SERVICE_VERSION || '0.1.0'

    const resource = new Resource({
        [ATTR_SERVICE_NAME]: 'omni-web',
        [ATTR_SERVICE_VERSION]: serviceVersion,
        'deployment.environment': environment,
        'deployment.id': deploymentId,
    })

    const traceExporter = otlpEndpoint
        ? new OTLPTraceExporter({
              url: `${otlpEndpoint}/v1/traces`,
          })
        : undefined

    sdk = new NodeSDK({
        resource,
        traceExporter,
        instrumentations: [
            getNodeAutoInstrumentations({
                '@opentelemetry/instrumentation-fs': {
                    enabled: false,
                },
            }),
        ],
    })

    sdk.start()

    if (otlpEndpoint) {
        console.log(`OpenTelemetry initialized with OTLP endpoint: ${otlpEndpoint}`)
    } else {
        console.log('No OTLP endpoint configured, telemetry will be collected locally only')
    }

    console.log(
        `Telemetry initialized for omni-web (deployment_id=${deploymentId}, environment=${environment})`,
    )

    // Graceful shutdown
    process.on('SIGTERM', async () => {
        try {
            await sdk?.shutdown()
            console.log('Telemetry shut down successfully')
        } catch (error) {
            console.error('Error shutting down telemetry', error)
        } finally {
            process.exit(0)
        }
    })
}

export function getTracer(name: string = 'omni-web') {
    return trace.getTracer(name)
}

export function injectTraceContext(headers: Record<string, string>): Record<string, string> {
    const activeContext = context.active()
    const carrier: Record<string, string> = { ...headers }

    propagation.inject(activeContext, carrier)

    return carrier
}

export function extractTraceContext(headers: Record<string, string | undefined>) {
    const carrier: Record<string, string> = {}

    for (const [key, value] of Object.entries(headers)) {
        if (value !== undefined) {
            carrier[key] = value
        }
    }

    return propagation.extract(context.active(), carrier)
}

export function getRequestId(): string | undefined {
    const span = trace.getActiveSpan()
    if (span) {
        return span.spanContext().traceId
    }
    return undefined
}

export function startSpan(name: string, fn: (span: Span) => Promise<any>) {
    const tracer = getTracer()
    return tracer.startActiveSpan(name, async (span) => {
        try {
            const result = await fn(span)
            span.end()
            return result
        } catch (error) {
            span.recordException(error as Error)
            span.end()
            throw error
        }
    })
}
